src/class-controller.php

```diff
@@ -0,0 +1,215 @@
+<?php
+/**
+ * Controller class file
+ *
+ * @package wp_theme_migrator
+ */
+
+namespace Alley\WP\Theme_Migrator;
+
+use Alley\WP\Theme_Migrator\Context;
+use Alley\WP\Theme_Migrator\WP as WP_Theme_Migrator_WP;
+use Exception;
+
+/**
+ * Controller class.
+ */
+class Controller {
+	use \Alley\WP\Theme_Migrator\Theme_Helpers;
+
+	/**
+	 * Context instance.
+	 *
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * WP_Theme_Migrator_WP instance.
+	 *
+	 * @var WP_Theme_Migrator_WP
+	 */
+	protected $wp;
+
+	/**
+	 * Whether to handle the current request with the theme we're migrating to.
+	 *
+	 * @var bool
+	 */
+	protected $should_migrate;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Context $context Context object.
+	 *
+	 * @throws Exception Thrown if context is not valid.
+	 */
+	public function __construct( Context $context ) {
+		if ( ! $context->is_valid_context() ) {
+			throw new Exception(
+				// @todo Add more robust error messaging.
+				sprintf(
+					__( 'WP Theme Migrator failed. Context is not valid.', 'wp-theme-migrator' ),
+				)
+			);
+		}
+
+		$this->context = $context;
+		$this->wp      = new WP_Theme_Migrator_WP();
+		$this->init();
+	}
+
+	/**
+	 * Add hooks to kick off migrator and tie in useful query vars.
+	 */
+	protected function init() {
+		// Post types and taxonomies must be registered by this point to be used
+		// for determining migratability.
+		add_action( 'setup_theme', [ $this, 'run' ], 100 );
+
+		// Add query vars for post types to our WP.
+		add_action( 'registered_post_type', [ $this, 'add_post_type_query_vars' ], 10, 2 );
+
+		// @todo Add query vars for taxonomies.
+	}
+
+	/**
+	 * Run the migrator.
+	 */
+	public function run() {
+		// Set up our extended WP instance to establish query vars early.
+		$this->set_up_wp();
+
+		// Determine migratability of the current request.
+		$this->set_migratability();
+
+		// Maybe migrate.
+		$this->maybe_migrate();
+	}
+
+	/**
+	 * Sets up WP_Theme_Migrator_WP instance and parses the request.
+	 */
+	protected function set_up_wp() {
+		$this->wp->main();
+	}
+
+	/**
+	 * Set migratability of the current request.
+	 */
+	protected function set_migratability() {
+		/**
+		 * Skip migration checks while doing cron.
+		 *
+		 * @param bool              $skip_during_cron Whether to skip.
+		 * @param WP_Theme_Migrator $this This WP_Theme_Migrator instance.
+		 */
+		$skip_during_cron = apply_filters( 'wp_theme_migrator_skip_during_cron', true, $this );
+
+		if ( $skip_during_cron && defined( 'DOING_CRON' ) && DOING_CRON ) {
+			return;
+		}
```

Comment:
I'd recommend doing a simple cron check before running the filter to avoid having the filter run on every request:
```suggestion
		if ( defined( 'DOING_CRON' ) && DOING_CRON ) {
			/**
			 * Skip migration checks while doing cron.
			 *
			 * @param bool              $skip_during_cron Whether to skip.
			 * @param WP_Theme_Migrator $this This WP_Theme_Migrator instance.
			 */
			if ( apply_filters( 'wp_theme_migrator_skip_during_cron', true, $this ) ) {
				return;
			}
		}
```